src/internal/collection/postgres_collection.go

```diff
@@ -301,80 +330,131 @@ func (c *postgresCollection) list(
 		query += " where " + strings.Join(fields, " and ")
 	}
 
-	if opts.Order != SortNone {
-		if order, err := orderToSQL(opts.Order); err != nil {
-			return err
-		} else if target, err := targetToSQL(opts.Target); err != nil {
-			return err
+	afterCondition := func(t SortTarget, ord SortOrder) (string, error) {
+		ts, err := targetToSQL(t)
+		if err != nil {
+			return "", err
+		}
+		afterSortVal, err := sortTargetValue(last, t)
+		if err != nil {
+			return "", err
+		}
+		args = append(args, afterSortVal)
+		// condition ends up including:
+		// where ... (<TARGET> <OP> <VALUE> or (<TARGET> = <VALUE> and key > '<PKEY>'))
+		// This covers the case where we only want to return rows following a certain offset value.
+		// We handle the case where multiple rows have the same sort value (when it isn't sorted by primary key)
+		// by including "or (<TARGET> = <VALUE> and key > '<PKEY>')" and also ordering by Primary Key as a tie breaker
+		cond := fmt.Sprintf("(%s %s $%d or (%s = $%d and key > '%s'))", ts, sortOrderOperator(ord), len(args), ts, len(args), last.Key)
+		if len(withFields) > 0 {
+			return " and " + cond, nil
+		} else {
+			return " where " + cond, nil
+		}
+	}
+
+	st := opts.Target
+	so := opts.Order
+	if opts.Order == SortNone { // defaults if to "order by key asc" if no sort specified
+		st = SortByKey
+		so = SortAscend
+	}
+	if order, err := orderToSQL(so); err != nil {
+		return "", nil, err
+	} else if target, err := targetToSQL(st); err != nil {
+		return "", nil, err
+	} else {
+		if last != nil {
+			ac, err := afterCondition(st, so)
+			if err != nil {
+				return "", nil, err
+			}
+			query += ac
+		}
+		if st == SortByKey {
+			query += fmt.Sprintf(" order by key %s", order)
 		} else {
-			query += fmt.Sprintf(" order by %s %s", target, order)
+			query += fmt.Sprintf(" order by %s %s, key asc", target, order)
 		}
 	}
 
 	if opts.Limit > 0 {
 		query += fmt.Sprintf(" limit %d", opts.Limit)
 	}
+	return query, args, nil
+}
 
-	rows, err := q.QueryxContext(ctx, query, args...)
-	if err != nil {
-		return c.mapSQLError(err, "")
-	}
-	defer rows.Close()
+func (c *postgresCollection) list(
+	ctx context.Context,
+	withFields map[string]string,
+	opts *Options,
+	q sqlx.ExtContext,
+	f func(*model) error,
+) error {
+	// To avoid holding a transaction open (which holds a DB connection) for an unknown duration
+	// dictated by the client's callback, we:
+	// (1) buffer the SQL rows
+	// (2) close the connection
```

Comment:
Close the transaction. We don't really explicitly do this as much as read a subset of the query results. Switching to the limit approach may warrant rewording this as performing a query with a limit to get a batch.